### ISSUE TITLE
Terminology: next round

### ISSUE THREAD

**Issue by cabo** [ietf-wg-jsonpath/draft-ietf-jsonpath-base#84]
Clusters of terms:

* Cluster 1

  * JSON value
  * (data) item
  * node
  * location

  These are related, but different.  We probably do not need all four.
  (Of course, JSON value is a 8259 term, so it does not go away.)
  We should decide which of these terms we want to use, and what they exactly mean.
  A node seems to be a combination of a discrete JSON value and its location.
  JSON value seems to imply equivalence by value (even though equivalence is undefined in RFC 8259),
  while node is based on equivalence by identity.
  Location seems to imply that you can get to the node (and thus the value) from it.

* Cluster 2

  * Argument: input to executing JSON query
  * root value, root node (used as Argument, may be used in nested queries)
  * current node (current item):
    * input to selector function invocation, 
    * maybe used inside filtering/indexing selectors

* Cluster 3

  * selector (part of the selector chain that makes up a JSONPath query)
  * (we probably need a word for) selector function \[invocation]

    This is needed if we stick to the model that each node in the
    input node list is treated separately and the results are then
    combined using (TBD) combine1.

  * combine1 (placeholder for what combines the outputs of the selector function invocations)

    This becomes important when we either resolve there is only one
    combine1 function or we resolve that combine1 functions are
    (within limits) implemenation specific (i.e., we give up) or that
    different selectors have different combine1 functions.

  All these can only be meaningfully decided once we have agreement on a processing model.

**Comment by goessner** [comment]
... (we probably need a word for) selector function [invocation] ...

perhaps ...
* selection
* selection process 
* selection step (atomic)

**Comment by danielaparker** [comment]
> Clusters of terms:
> 
> * Cluster 1
>   
>   * JSON value
>   * (data) item
>   * node
>   * location
>   
>   These are related, but different.  We probably do not need all four.
>   (Of course, JSON value is a 8259 term, so it does not go away.)

My view is that "JSON value" is enough, that "data item" is unnecessary  

>   We should decide which of these terms we want to use, and what they exactly mean.
>   A node seems to be a combination of a discrete JSON value and its location.

Agreed

>   JSON value seems to imply equivalence by value (even though equivalence is undefined in RFC 8259),
>   while node is based on equivalence by identity.
>   Location seems to imply that you can get to the node (and thus the value) from it.

Agreed

> * Cluster 2
>   
>   * Argument: input to executing JSON query
>   * root value, root node (used as Argument, may be used in nested queries)

My view is that Argument is too generic a term, e.g. it is common to talk about providing an argument to a function. I like @timbray 's suggestion of root value. "root" is used in Goessner 2007.  

>   * current node (current item):
>     
I like "current node", with "node" as understood above

>     * input to selector function invocation,
>     * maybe used inside filtering/indexing selectors



> * Cluster 3
>   
>   * selector (part of the selector chain that makes up a JSONPath query)
>   * (we probably need a word for) selector function [invocation]
>     This is needed if we stick to the model that each node in the
>     input node list is treated separately and the results are then
>     combined using (TBD) combine1.
>   * combine1 (placeholder for what combines the outputs of the selector function invocations)
>     This becomes important when we either resolve there is only one
>     combine1 function or we resolve that combine1 functions are
>     (within limits) implementation specific (i.e., we give up) or that
>     different selectors have different combine1 functions.
>   
>   All these can only be meaningfully decided once we have agreement on a processing model.

The main thing is that the "processing model" section is shorter and simpler than the one in [XPATH 3.1 Processing Model](https://www.w3.org/TR/xpath-31/#id-processing-model) :-) The term "processing model" frightens me because of this association.

There are different ways of describing the evaluation of an expression. The one in the PR considers each selector operating on an input nodelist and producing an output nodelist. I think if I understand it correctly, that's consistent with how the XPath 3.1 spec describes evaluation.

Another approach would be to consider an abstract select function that takes the current value and the right expression (the "." or "[" and everything to the right) and produces a sequence of expressions,  each of which are recursively evaluated, until arriving at values or no values.   For example:

Example 1

Given a root value `{"foo": {"bar": "baz"}}`, and a JSONPath expression: `$foo.bar`
```
select(".foo", {"foo": {"bar": "baz"}}) -> {"bar": "baz"}
```
Example 2

Given a root value `{"a":[{"b":0},{"b":1},{"c":2}]}` and the JSONPath `$.a[*].b`
```
select(".a[*].b", {"foo": {"a":[{"b":0},{"b":1},{"c":2}]}})  

-> select("[*].b", [{"b":0},{"b":1},{"c":2}])

-> select(".b", {"b":0}) -> 0, select(".b", {"b":1}) -> 1, select(".b", {"b":2}) -> 2
 ```
This is kind of how JMESPath describes JMESPath evaluation, adapted to returning sequence of results. 
.
Mentioned for comparative purposes only, I have no opinion.

**Comment by cabo** [comment]
> On 2021-03-27, at 03:12, Daniel Parker ***@***.***> wrote:
> 
> My view is that Argument is too generic a term, e.g. it is common to talk about providing an argument to a function. I like @timbray 's suggestion of root value. "root" is used in Goessner 2007.

We are using way more generic terms: value, object etc.
(I’d rather get rid of those first.)

The Argument is the JSON value that goes into the evaluation of the query.
The Argument is then viewed as a tree and *has* a root node.

Grüße, Carsten

**Comment by goessner** [comment]
... using some examples ...

* `$.obj['city']` ... `'city'` is the *name* of a member value `"Berlin"` of object `obj`. 
* `$.arr[12]` ... `12` is the *index* of an entry value `3.14` of array `arr`. 
* *location* is a general/abstract term for *name* and *index*. So `'city'` and `12` are both locations.
* `$` is *name* and *location* of the root value.
* A (normalized) path is a sequence of locations enclosed in brackets.
* A `node` is a `location / value` pair.
* A `selector` in its simpliest case is a location. Special selectors are `*, .., @`.

I think with this we have a minimalistic and complete term set.

**Comment by danielaparker** [comment]
> ... using some examples ...
> 
> * _location_ is a general/abstract term for _name_ and _index_. So `'city'` and `12` are both locations.

I think a _location_ is an abstraction that represents a full path, it has to uniquely identify a JSON value within a root value. In terms of its actual representation, it could be a normalized path, or an array of name or index components, or a name or index component with a parent.

> * A (normalized) path is a sequence of locations enclosed in brackets.

I think a normalized path is a property of a location.

**Comment by glyn** [comment]
Unless I've missed a strict definition in our discussions, there can be more than one normalised path for a given location.

For instance, to identify the child node of the root node with name `a`, and assuming we are using the more general "bracket" form of child name (the `child` production in the current grammar), we could use an array consisting of any one of the following name components:
* "a"
* 'a'
* "u0061"
* '\u0061'

To accommodate this, we could associate a location with an [equivalence class](https://en.wikipedia.org/wiki/Equivalence_class) of normalised paths.

Alternatively, we could define a unique normalised path for each location (e.g. for names by stipulating double quotes and unicode escapes for every character, even if that would be unfriendly to end users and, for indices, by stipulating non-negative indices). I guess we could even standardise on basic Latin unicode characters and use unicode escapes for everything else, but that seems a little biased.

**Comment by cabo** [comment]
> * "a"
> * 'a'
> * "u0061"
> * '\u0061'

That is not exactly a terminology issue.

If we want the output paths to be normalized, we would need to define that normalization.
If we just want them to be stylized (e.g., bracket notation only, unsigned integer indexing or quoted name indexing) and leave representation details to the implementations, we could do that, too.

(My personal view is that they should be normalized.  Single quotes are very un-JSON, and escaping non-control characters is un-2000s, so it is pretty clear that "a" would be the normalized form.)

**Comment by danielaparker** [comment]
> Unless I've missed a strict definition in our discussions, there can be more than one normalised path for a given location.
> 
> For instance, to identify the child node of the root node with name `a`, and assuming we are using the more general "bracket" form of child name (the `child` production in the current grammar), we could use an array consisting of any one of the following name components:
> 
> * "a"
> * 'a'
> * "u0061"
> * '\u0061'
> 
> To accommodate this, we could associate a location with an [equivalence class](https://en.wikipedia.org/wiki/Equivalence_class) of normalised paths.
> 
> Alternatively, we could define a unique normalised path for each location (e.g. for names by stipulating double quotes and unicode escapes for every character, even if that would be unfriendly to end users and, for indices, by stipulating non-negative indices). I guess we could even standardise on basic Latin unicode characters and use unicode escapes for everything else, but that seems a little biased.

To be consistent with Goessner 20007, and existing practice, I think it needs to be defined in terms of bracket notation with only non-negative indices and single quoted names allowed, and escaping required if the name includes a single quote character. But yes that still leaves an issue with uXXXX.

**Comment by cabo** [comment]
> @goessner wrote:
> 
> > * _location_ is a general/abstract term for _name_ and _index_. So `'city'` and `12` are both locations.
> 
> I think a _location_ is an abstraction that represents a full path, 

I think that works best the other way around: a location is an abstraction that can be represented by a path (full path, path from the root, ...).

> it has to uniquely identify a JSON value within a root value. 

We don't really need locations for member names, so I would limit this to nodes.

> In terms of its actual representation, it could be a normalized path, or an array of name or index components, or a name or index component with a parent.
> 
> > * A (normalized) path is a sequence of locations enclosed in brackets.
> 
> I think a _normalized path_ is a property of a location.

If we have normalization of paths, a normalized path can indeed be derived from the path (from the root) of the location.  (But I don't know yet whether we opt for normalization of output paths.)

**Comment by goessner** [comment]
*  charming aspect of single quotes in normalized pathes would be to use them as JSON values itself without escaping ...
* We should care for: There can be only one normalized path !
* If `location` is defined as the complete path, we need another universal term for `name / index`.

**Comment by timbray** [comment]
I really, really like Stefan's approach here. I it is precise and easy to
understand and does the minimum amount of invention required.

**Comment by timbray** [comment]
On Sat, Mar 27, 2021 at 4:41 AM Glyn Normington ***@***.***>
wrote:

> Unless I've missed a strict definition in our discussions, there can be
> more than one normalised path for a given location.
>
> For instance, to identify the child node of the root node with name a,
> and assuming we are using the more general "bracket" form of child name
> (the child production in the current grammar), we could use an array
> consisting of any one of the following name components:
>
>    - "a"
>    - 'a'
>    - "u0061"
>    - '\u0061'
>
> While I normally run screaming in the other direction when people propose
a "canonical form" of anything, I think we could get away with it here. Any
JSON string can be represented as a sequence of Unicode codepoints -
granted, some of them might not be *valid* codepoints - and for the purpose
of a normalized path, we could either (a) specify UTF-8 or (b) say that the
codepoints are all that matters, the syntactic expression is not
significant.

Or am I missing something?

**Comment by danielaparker** [comment]
@goessner wrote: 
 
> hmm ... why not simply use `path` instead of `location` from root to value and redefine `location` as local location relative to parent value ... ?

My understanding was that a more abstract term than "path" was desired, but I'll defer to @cabo  and @glyn.

**Comment by glyn** [comment]
> hmm ... why not simply use `path` instead of `location` from root to value

I don't feel strongly. I think we could define an abstract "path" term, but I wonder if it might sound a bit odd in places in context of _JSONpath_?

> and redefine `location` as local location relative to parent value ... ?

I'm not sure we need a term for that. We already have names in objects and indices in arrays.

**Comment by cabo** [comment]
It would help the concision to have a single common term for indexing by name and indexing by index.  I don't have a proposal, though...
("location" probably is not it, that term is better used the way it is now.  A location in that sense can be expressed by an output path, but it "is" not an output path.)

**Comment by gregsdennis** [comment]
> It would help the concision to have a single common term for indexing by name and indexing by index. - @cabo 

This has been brought up before.  To me, `child` is the only sensible term that covers both objects and arrays.  I know there's other discussion regarding this term, but I'm on mobile and can't find it right now.

**Comment by cabo** [comment]
"child" gives me the XPath heebie-jeebies...  But that is a reasonable working term for array element or member value (i.e., things that result from indexing a container).

What's missing is a term for the parameter that goes into the indexing (array index or member name).  Maybe we need to rename array index so we can use index for the combined term (which also would solve the fact that indexing otherwise isn't always related to indexes).

**Comment by gregsdennis** [comment]
> Maybe we need to rename array index so we can use index for the combined term (which also would solve the fact that indexing otherwise isn't always related to indexes).

This works for me. I tend to think of a thing inside `[]` as an index, whether it's a number or a string (because the `[]` in .Net is called the "index" operator).

It also helps define unions as "a set of indices" where a single "index" could be a number, a string, a slice, etc. Unions can then be any combination of these things.

(We should also decide on "indices" vs "indexes". I'm not sure which is more correct.)

**Comment by goessner** [comment]
Search engine of my choice offers "tag, label, mark" as kind of synonyms for "name / index". So how do we like to talk about:

* tag/value pair
* label/value pair
* mark/value pair

"tag" may be too HTML'ish ... ?

**Comment by gregsdennis** [comment]
Given those options, I think I'll stick with my vote for "index".

**Comment by goessner** [comment]
Don't forget "key" ...

* key/value pair

**Comment by cabo** [comment]
I think we can stick with 8259 for member name.  Key is maybe indeed a candidate for "member name or array index".

I'll propose a PR as soon as #90 is through (too many conflicts otherwise).

**Comment by goessner** [comment]
Following on from the previous comments, I propose to add two additional terms to the terminology list:
* **Key**: Synonym for array index or object member name.

* **Selector**: An atomic component of a query expression. Different selector types are ...
    * Root selector `'$'`
    * Current item selector `'@'`
    * Nested descendants selector `'..'`.
    * Dot selector `'.<name>'`, used with object member names exclusively.
    * Dot wild card selector `'.*'`.
    * Key selector `'[<key>]'`, where `<key>` is either an (possibly negative) array index or an object member name.
    * Key wild card selector `'[*]'`.
    * Array slice selector `'[<start>:<end>:<step>]'`, where `<start>`, `<end>`, `<step>` are integer literals.
    * Union selector `'[<key1>,<key2>,...,<keyN>]'`, holding a comma delimited list of keys.
    * Filter selector `[?(<expr>)]`

I hope I did not miss a selector type. Also, I dislike the current fact, that union selectors are first class citizens in the spec. They should be equal among equal. Let's discuss ...

**Comment by cabo** [comment]
On 2021-04-29, at 16:39, Stefan Goessner ***@***.***> wrote:
> 
> 	• Key: Synonym for array index or object member name.

How about maybe index/key?  index key?  indexing key?

Grüße, Carsten

**Comment by goessner** [comment]
We will use that mostly in the form "key/value pair", where "index/key/value pair" is confusing. "index/value pair" would be better then, but "index" alone is often mentally completed to "array index".

**Comment by goessner** [comment]
... if agreement goes to "index/value pair", I can live with it ...

**Comment by gregsdennis** [comment]
I still think "index" demoted an array specifically.

**Comment by goessner** [comment]
ok ... then ...

* **Index**: Synonym for array index or object member name.

* **Selector**: An atomic component of a query expression. Different selector types are ...
    * Root selector `'$'`
    * Current item selector `'@'`
    * Nested descendants selector `'..'`.
    * Dot selector `'.<name>'`, used with object member names exclusively.
    * Dot wild card selector `'.*'`.
    * Index selector `'[<index>]'`, where `<index>` is either an (possibly negative) array index or an object member name.
    * Index wild card selector `'[*]'`.
    * Array slice selector `'[<start>:<end>:<step>]'`, where `<start>`, `<end>`, `<step>` are integer literals.
    * Union selector `'[<index1>,<index2>,...,<indexN>]'`, holding a comma delimited list of indices (indexes ?).
    * Filter selector `[?(<expr>)]`

**Comment by gregsdennis** [comment]
"Indices" is more correct, but "indexes" has become more common recently.

Also, I still think we should include the `[(<expr>)]`  selector.  I've already explained its uses in other issues, but we should keep it for now other reason than people will complain if it's not supported.

**Comment by goessner** [comment]
Well ... I would like to get filter selector syntax straight first. Then discuss, if and what subset of it we want to support with index selectors.

**Comment by gregsdennis** [comment]
I would expect that the expression syntax would be the same. The resulting value would need to, instead of being truthy, be able to evaluate to a well-defined index, e.g. a number or string.

**Comment by cabo** [comment]
>   * Index selector `'[<index>]'`, where `<index>` is either an (possibly negative) array index or an object member name.
>   * Array slice selector `'[<start>:<end>:<step>]'`, where `<start>`, `<end>`, `<step>` are integer literals.
>   * Union selector `'[<index1>,<index2>,...,<indexN>]'`, holding a comma delimited list of indices (indexes ?).

I think we should summarize all the above kinds of selector into "indexing selectors".
(Note that the wild card selector is not there, as there is no indexing key.)

**Comment by mkmik** [comment]
The current draf says:
> union-element =/ array-index / array-slice

Thus a an indexing expression can include multiple slice expressions:

`'[<start1>:<end1>:<step1>,<start2>:<end2>:<step2>]'

**Comment by cabo** [comment]
Closed by #97 (but "resolves" doesn't trigger closing the issue, so I'm doing this by hand here).
